src/libsystemd/sd-id128/sd-id128.c

```diff
@@ -181,3 +182,34 @@ _public_ int sd_id128_randomize(sd_id128_t *ret) {
         *ret = make_v4_uuid(t);
         return 0;
 }
+
+_public_ int sd_id128_get_machine_app_specific(sd_id128_t app_id, sd_id128_t *ret) {
+        _cleanup_(sd_hash_unrefp) sd_hash *h = NULL;
+        sd_id128_t m, result;
+        const void *p;
+        int r;
+
+        assert_return(ret, -EINVAL);
+
+        r = sd_id128_get_machine(&m);
+        if (r < 0)
+                return r;
+
+        r = sd_hash_new_with_key(&h, "hmac(sha256)", &m, sizeof(m));
+        if (r < 0)
+                return r;
+
+        r = sd_hash_put(h, &app_id, sizeof(app_id));
+        if (r < 0)
+                return r;
+
+        r = sd_hash_digest_data(h, &p);
+        if (r < 0)
+                return r;
+
+        /* We chop off the trailing 16 bytes */
+        memcpy(&result, p, MIN(sd_hash_get_size(h), sizeof(result)));
```

Comment:
sha256 is a 256bit hash, but we want to return an uuid-like object, and uuids generally have 128bit. hence we simply chop off 128bit.

of course, we could also use hmac-sha1, but sha1 isn't recommended anymore.